src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -1002,24 +1004,44 @@ internal static object ReplaceOperator(ExecutionContext context, IScriptExtent e
         /// <param name="input">The input string.</param>
         /// <param name="regex">A Regex instance.</param>
         /// <param name="substitute">The substitute value.</param>
+        /// <param name="cachedReplacementString">Cached replacement string.</param>
+        /// <param name="cachedMatchEvaluator">Cached MatchEvaluator.</param>
         /// <returns>The result of the regex.Replace operation.</returns>
-        private static object ReplaceOperatorImpl(ExecutionContext context, string input, Regex regex, object substitute)
+        private static object ReplaceOperatorImpl(
+            ExecutionContext context,
+            string input,
+            Regex regex,
+            object substitute,
+            ref string cachedReplacementString,
+            ref MatchEvaluator cachedMatchEvaluator)
         {
+            // After first call we will have necessarily either cachedReplacementString or cachedMatchEvaluator.
+            if (cachedReplacementString is not null)
+            {
+                return regex.Replace(input, cachedReplacementString);
+            }
+
+            if (cachedMatchEvaluator is not null)
+            {
+                return regex.Replace(input, cachedMatchEvaluator);
+            }
+
             switch (substitute)
             {
-                case string replacementString:
-                    return regex.Replace(input, replacementString);
+                case string replacement:
+                    cachedReplacementString = replacement;
+                    return regex.Replace(input, cachedReplacementString);
 
                 case ScriptBlock sb:
-                    MatchEvaluator me = GetMatchEvaluator(context, sb);
-                    return regex.Replace(input, me);
+                    cachedMatchEvaluator = GetMatchEvaluator(context, sb);
+                    return regex.Replace(input, cachedMatchEvaluator);
 
-                case object val when LanguagePrimitives.TryConvertTo(val, out MatchEvaluator matchEvaluator):
-                    return regex.Replace(input, matchEvaluator);
+                case object val when LanguagePrimitives.TryConvertTo(val, out cachedMatchEvaluator):
+                    return regex.Replace(input, cachedMatchEvaluator);
 
                 default:
-                    string replacement = PSObject.ToStringParser(context, substitute);
-                    return regex.Replace(input, replacement);
+                    cachedReplacementString = PSObject.ToStringParser(context, substitute);
+                    return regex.Replace(input, cachedReplacementString);
```

Comment:
It would be better if we can reduce the `return statements`. How about the following?
```c#
            if (cachedReplacementString is null && cachedMatchEvaluator is null)
            {
                switch (substitute)
                {
                    case string replacement:
                        cachedReplacementString = replacement;
                        break;

                    case ScriptBlock sb:
                        cachedMatchEvaluator = GetMatchEvaluator(context, sb);
                        break;

                    case object val when LanguagePrimitives.TryConvertTo(val, out cachedMatchEvaluator):
                        break;

                    default:
                        cachedReplacementString = PSObject.ToStringParser(context, substitute);
                        break;
                }
            }

            if (cachedReplacementString is not null)
            {
                return regex.Replace(input, cachedReplacementString);
            }
            else
            {
                return regex.Replace(input, cachedMatchEvaluator);
            }
```

Reply:
> It would be better

Do you say about readability or anything else? 

Readability with return looks good for me - it _explicitly_ shows design intensions.
.Net generates optimal code https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdACpwAeM1ASnAOYCuANhAgKIMADgjhQoASwD2AOygBuYrnLxYxAN7Fy2ndqwBmSngBs5KcABWcAMYxybIb2twA8kMQQYUhAEkwjgAotXRC9PAoJGSFOGCRg0N02dkZyUWSGOKIEkPMrW3IoTmBYCRgYuEzs3VEAMyMKawhrAAs4ABMHJzgwOBkYAGUYBEj2SqrtWvIAWU8W/gA3CG5OT29yRpb2mZg5xeXVhABKeJDNLPHyAHpL8gBBGvgEchqJBFh1pe5yAHc4H4luF9mhB5n8ZDYxFA+ACAJ7kOClVpPDatDpwRxNbq9AZDEZmZFNVHbXZLFZeBDUE7ZCR1AIo9qdTE9PqDYYydjkCRQcgyKR2GQ8bjHc7jM4XEJYADsqQ4jFY6K6AUi0ViH02aIxzmZOLZ7EOihFVQAvsQqQkaeQ6YStrNmgtSQdOdzefzBcLxWLxTopTL0vLNXAlVEYkg1UTbfb9uT9WbQiaiLGQlBvqUWpbCsUYKVyu6Lp6vR8oH8qBRRAHtSBE+L6RqutrWXiALwy8vYg0F0I+tJyxnOIMq0M13tYlm49kxhOG6vQP79azDIQwABC3Ck1gA1gVQFWLkOFUzsQ32eRmwAiOcL5erjen9sd3Rd2UMf2K5UhsM2nZ2vZk7wTnehI0RZmJYNh2HsPytDI5AQVyPJ8jygqVlOHZ7q2o66ie5CnrkYG3gBVSPn6w79u+NbEt+DrRgaBG6G0cA1BAPCoLRCRoXWh5jhyZ70YxzH4ShBZET2+59m+qrsQeGEjBO4rxtk8bxkoODkBeEiLiua7rhoRpAA===

Reply:
Or 
```csharp
            if (cachedReplacementString is null && cachedMatchEvaluator is null)
            {
                (cachedReplacementString, cachedMatchEvaluator) = substitute switch
                {
                    string replacement => (replacement, cachedMatchEvaluator),
                    ScriptBlock sb => (cachedReplacementString, GetMatchEvaluator(context, sb)),
                    { } val when LanguagePrimitives.TryConvertTo(val, out cachedMatchEvaluator) => (cachedReplacementString, cachedMatchEvaluator),
                    _ => (PSObject.ToStringParser(context, substitute), cachedMatchEvaluator),
                };
            }
            // After first call we will have necessarily either cachedReplacementString or cachedMatchEvaluator.
            return cachedReplacementString is { } replacementString
                ? regex.Replace(input, replacementString)
                : regex.Replace(input, cachedMatchEvaluator);
```